### ISSUE TITLE
Adjust all pending active fields in yas-expand-snippet

### ISSUE THREAD

**Issue by netjune** [joaotavora/yasnippet#1038]
Currently, only one field is adjusted in yas-expand-snippet. It causes problem for stacked snippets with level be lager than 2.

I set yas-verbose to 4 and add code to print the active snippets in yas--message:
```
(defun yas--message (level message &rest args)
  "When LEVEL is at or below `yas-verbosity', log MESSAGE and ARGS."
  (when (>= yas-verbosity level)
    (message "%s" (apply #'yas--format message args))
    (message "--- active snippets: %s\n" yas--active-snippets)
    ))
```

There are 3 nested snippets. The log shows that one middle snippet(the snippet id: 1) leaked.

The log content:
```
[yas] snippet 0 expanded.
--- active snippets: (#s(yas--snippet nil (#s(yas--field 1 #<marker at 14989 in device_center.go> #<marker at 14998 in device_center.go> nil nil nil nil #s(yas--exit #<marker at 15002 in device_center.go> nil))) #s(yas--exit #<marker at 15002 in device_center.go> nil) 0 #<overlay from 14974 to 15005 in device_center.go> #s(yas--field 1 #<marker at 14989 in device_center.go> #<marker at 14998 in device_center.go> nil nil nil nil #s(yas--exit #<marker at 15002 in device_center.go> nil)) nil nil))

[yas] snippet 1 expanded.
--- active snippets: (#s(yas--snippet nil (#s(yas--field 1 #<marker at 14991 in device_center.go> #<marker at 14993 in device_center.go> nil nil nil nil nil)) nil 1 #<overlay from 14989 to 14993 in device_center.go> #s(yas--field 1 #<marker at 14991 in device_center.go> #<marker at 14993 in device_center.go> nil nil nil nil nil) #s(yas--field 1 #<marker at 14989 in device_center.go> #<marker at 14993 in device_center.go> nil nil nil t #s(yas--exit #<marker at 14997 in device_center.go> nil)) nil) #s(yas--snippet nil (#s(yas--field 1 #<marker at 14989 in device_center.go> #<marker at 14993 in device_center.go> nil nil nil t #s(yas--exit #<marker at 14997 in device_center.go> nil))) #s(yas--exit #<marker at 14997 in device_center.go> nil) 0 #<overlay from 14974 to 15000 in device_center.go> #s(yas--field 1 #<marker at 14989 in device_center.go> #<marker at 14993 in device_center.go> nil nil nil t #s(yas--exit #<marker at 14997 in device_center.go> nil)) nil nil))

[yas] snippet 2 expanded.
--- active snippets: (#s(yas--snippet nil nil nil 2 #<overlay from 14996 to 15010 in device_center.go> nil #s(yas--field 1 #<marker at 14991 in device_center.go> #<marker at 15010 in device_center.go> nil nil nil t nil) t) #s(yas--snippet nil (#s(yas--field 1 #<marker at 14991 in device_center.go> #<marker at 15010 in device_center.go> nil nil nil t nil)) nil 1 #<overlay from 14989 to 15010 in device_center.go> #s(yas--field 1 #<marker at 14991 in device_center.go> #<marker at 15010 in device_center.go> nil nil nil t nil) #s(yas--field 1 #<marker at 14989 in device_center.go> #<marker at 14996 in device_center.go> nil nil nil t #s(yas--exit #<marker at 15014 in device_center.go> nil)) nil) #s(yas--snippet nil (#s(yas--field 1 #<marker at 14989 in device_center.go> #<marker at 14996 in device_center.go> nil nil nil t #s(yas--exit #<marker at 15014 in device_center.go> nil))) #s(yas--exit #<marker at 15014 in device_center.go> nil) 0 #<overlay from 14974 to 15017 in device_center.go> #s(yas--field 1 #<marker at 14989 in device_center.go> #<marker at 14996 in device_center.go> nil nil nil t #s(yas--exit #<marker at 15014 in device_center.go> nil)) nil nil))

[yas] Snippet 2 exited.
--- active snippets: (#s(yas--snippet nil (#s(yas--field 1 #<marker at 14991 in device_center.go> #<marker at 15010 in device_center.go> nil nil nil t nil)) nil 1 #<overlay from 14989 to 15010 in device_center.go> #s(yas--field 1 #<marker at 14991 in device_center.go> #<marker at 15010 in device_center.go> nil nil nil t nil) #s(yas--field 1 #<marker at 14989 in device_center.go> #<marker at 14996 in device_center.go> nil nil nil t #s(yas--exit #<marker at 15014 in device_center.go> nil)) nil) #s(yas--snippet nil (#s(yas--field 1 #<marker at 14989 in device_center.go> #<marker at 14996 in device_center.go> nil nil nil t #s(yas--exit #<marker at 15014 in device_center.go> nil))) #s(yas--exit #<marker at 15014 in device_center.go> nil) 0 #<overlay from 14974 to 15017 in device_center.go> #s(yas--field 1 #<marker at 14989 in device_center.go> #<marker at 14996 in device_center.go> nil nil nil t #s(yas--exit #<marker at 15014 in device_center.go> nil)) nil nil))

[yas] Snippet 0 exited.
--- active snippets: (#s(yas--snippet nil (#s(yas--field 1 #<marker at 14991 in device_center.go> #<marker at 15010 in device_center.go> nil nil nil t nil)) nil 1 #<overlay from 14989 to 15010 in device_center.go> #s(yas--field 1 #<marker at 14991 in device_center.go> #<marker at 15010 in device_center.go> nil nil nil t nil) #s(yas--field 1 14989 14996 nil nil nil t #s(yas--exit 15014 nil)) nil))

Error in post-command-hook (yas--post-command-handler): (wrong-type-argument markerp 14996)
```

**Comment by netjune** [comment]
It can be fixed by the following patch:
```
diff --git a/yasnippet.el b/yasnippet.el
index e0b5537..fef8c76 100644
--- a/yasnippet.el
+++ b/yasnippet.el
@@ -4110,7 +4110,16 @@ for normal snippets, and a list for command snippets)."
                                         (overlay-get yas--active-field-overlay 'yas--field))))
                (when existing-field
                  (setf (yas--snippet-previous-active-field snippet) existing-field)
-                 (yas--advance-end-maybe existing-field (overlay-end yas--active-field-overlay))))
+                 ;; (yas--advance-end-maybe existing-field (overlay-end yas--active-field-overlay))))
+                 ;; Adjust any pending active fields in case of stacked
+                 ;; expansion.
+                 (let* ((overlay yas--active-field-overlay)
+                        (psnippets (yas--gather-active-snippets
+                                    overlay start end t)))
+                   (dolist (psnippet psnippets)
+                     (let ((pfield (yas--snippet-previous-active-field psnippet)))
+                       (when pfield
+                         (yas--advance-end-maybe pfield (overlay-end overlay))))))))
 
              ;; Exit the snippet immediately if no fields.
              (unless (yas--snippet-fields snippet)
```

**Comment by npostavs** [comment]
Thanks for reporting, I think this should be fixed by #1044.
